### PR TITLE
Add GetMixins to mobilecoind API

### PR DESCRIPTION
**Pull request by mfaulk**
### Motivation

Adds API endpoint for getting TxOuts for use as mixins.

### In this PR
* GetMembershipProofRequest and GetMembershipProofResponse use TxOut instead of UnspentTxOut.
* Add GetMixins endpoint 
* Update mobilecoind-json
* TODO: Add unit tests to mobilecoind-json

### CODE REVIEW

File: mobilecoind/src/payments.rs [hunk 1]
```diff
@@ -228,6 +242,11 @@ impl<T: UserTxConnection + 'static, FPR: FogPubkeyResolver + Send + Sync + 'stat
         Ok(tx_proposal)
     }
 
+    /// Create a TxProposal that attempts to merge multiple UTXOs into a single larger UTXO.
+    ///
+    /// # Arguments
+    /// * `monitor_id` - ???
```
Comment (eranrund) [inline mobilecoind/src/payments.rs:248]: I believe this generates a transaction that joins a bunch of UTXOs belonging to monitor_id+subaddress_index

File: mobilecoind/src/payments.rs [hunk 2]
```diff
@@ -323,10 +345,17 @@ impl<T: UserTxConnection + 'static, FPR: FogPubkeyResolver + Send + Sync + 'stat
         Ok(tx_proposal)
     }
 
+    /// Create a TxProposal that sends the total value of all inputs minus the fee to a single receiver.
+    ///
+    /// # Arguments
+    /// * `account_key` -
```
Comment (eranrund) [inline mobilecoind/src/payments.rs:351]: The account key the UTXOs belong to

File: mobilecoind/src/payments.rs [hunk 3]
```diff
@@ -596,45 +620,57 @@ impl<T: UserTxConnection + 'static, FPR: FogPubkeyResolver + Send + Sync + 'stat
...
-    /// Build a TxProposal object.
+    /// Create a TxProposal.
+    ///
+    /// # Arguments
+    /// * `inputs` - UTXOs to spend, with membership proofs.
+    /// * `rings` - A set of mixins for each input, with membership proofs.
+    /// * `fee` - Transaction fee, in picoMOB.
+    /// * `from_account_key` - ???
+    /// * `change_subaddress` - ???
+    /// * `destinations` - ???
```
Comment (eranrund) [inline mobilecoind/src/payments.rs:669]: The outputs of the transaction

File: mobilecoind/src/payments.rs [hunk 4]
```diff
@@ -596,45 +620,57 @@ impl<T: UserTxConnection + 'static, FPR: FogPubkeyResolver + Send + Sync + 'stat
...
+    /// * `tombstone_block` - ???
```
Comment (eranrund) [inline mobilecoind/src/payments.rs:670]: Tombstone block of the transaction

File: mobilecoind/src/payments.rs [hunk 5]
```diff
@@ -596,45 +620,57 @@ impl<T: UserTxConnection + 'static, FPR: FogPubkeyResolver + Send + Sync + 'stat
...
+    /// * `fog_pubkey_resolver` - ???
```
Comment (eranrund) [inline mobilecoind/src/payments.rs:671]: Used to obtain fog report key when fog support is enabled.

File: mobilecoind/src/payments.rs [hunk 6]
```diff
@@ -596,45 +620,57 @@ impl<T: UserTxConnection + 'static, FPR: FogPubkeyResolver + Send + Sync + 'stat
...
     }
 
-    /// Build a TxProposal object.
+    /// Create a TxProposal.
+    ///
+    /// # Arguments
+    /// * `inputs` - UTXOs to spend, with membership proofs.
+    /// * `rings` - A set of mixins for each input, with membership proofs.
+    /// * `fee` - Transaction fee, in picoMOB.
+    /// * `from_account_key` - ???
```
Comment (eranrund) [inline mobilecoind/src/payments.rs:667]: Account key the inputs belong to and change is sent to

File: mobilecoind/src/payments.rs [hunk 7]
```diff
@@ -596,45 +620,57 @@ impl<T: UserTxConnection + 'static, FPR: FogPubkeyResolver + Send + Sync + 'stat
...
+    /// * `change_subaddress` - ???
```
Comment (eranrund) [inline mobilecoind/src/payments.rs:668]: Subaddress to send change to

File: mobilecoind/src/service.rs [hunk 8]
```diff
@@ -633,31 +636,35 @@ impl<
...
+        let outputs: Vec<TxOut> = request
+            .get_outputs()
             .iter()
-            .map(|proto_utxo| {
+            .map(|tx_out| {
                 // Proto -> Rust struct conversion.
-                UnspentTxOut::try_from(proto_utxo)
+                TxOut::try_from(tx_out)
                     .map_err(|err| rpc_internal_error("unspent_tx_out.try_from", err, &self.logger))
+                // UnspentTxOut::try_from(tx_out)
```
Comment (eranrund) [inline mobilecoind/src/service.rs:651]: Leftovers

**Review by eranrund — COMMENTED** [review]
Looks good so far!

File: mobilecoind-json/src/bin/main.rs [hunk 9]
```diff
@@ -677,6 +677,95 @@ fn tx_out_get_block_index_by_public_key(
...
+
+    let response = JsonMembershipProofResponse {
+        outputs,
+        membership_proofs,
+    };
+
+    Ok(Json(response))
+}
+
+#[post("/tx-out/mixin", format = "json", data = "<request>")]
```
Comment (eranrund) [inline mobilecoind-json/src/bin/main.rs:725]: mixins intead of mixin?

**Review by eranrund — APPROVED** [review]

**Comment by jadeydi** [comment]
May JsonBlockDetailsResponse return full data, like JsonTxOut, key images, etc?

https://github.com/mobilecoinfoundation/mobilecoin/blob/master/mobilecoind-json/src/data_types.rs#L1148

**Review by tsegaran — APPROVED** [review]

**Pull request merged**